src/Aspire.Dashboard/Model/DashboardClient.cs

```diff
@@ -0,0 +1,355 @@
+                                changes.Add(new(ResourceViewModelChangeType.Upsert, viewModel));
+                            }
+                        }
+                        else if (response.KindCase == WatchResourcesUpdate.KindOneofCase.Changes)
+                        {
+                            // Apply changes to the model.
+                            foreach (var change in response.Changes.Value)
+                            {
+                                changes ??= [];
+
+                                if (change.KindCase == WatchResourcesChange.KindOneofCase.Upsert)
+                                {
+                                    // Upsert (i.e. add or replace)
+                                    var viewModel = change.Upsert.ToViewModel();
+                                    _resourceByName[change.Upsert.Name] = viewModel;
+                                    changes.Add(new(ResourceViewModelChangeType.Upsert, viewModel));
+                                }
+                                else if (change.KindCase == WatchResourcesChange.KindOneofCase.Delete)
+                                {
+                                    // Remove
+                                    if (_resourceByName.Remove(change.Delete.ResourceName, out var removed))
+                                    {
+                                        changes.Add(new(ResourceViewModelChangeType.Delete, removed));
+                                    }
+                                    else
+                                    {
+                                        Debug.Fail("Attempt to remove an unknown resource view model.");
+                                    }
+                                }
+                                else
+                                {
+                                    throw new FormatException($"Unexpected {nameof(WatchResourcesChange)} kind: {change.KindCase}");
+                                }
+                            }
+                        }
+                        else
+                        {
+                            throw new FormatException($"Unexpected {nameof(WatchResourcesUpdate)} kind: {response.KindCase}");
+                        }
+                    }
+
+                    if (changes is not null)
+                    {
+                        foreach (var channel in _outgoingChannels)
+                        {
+                            // TODO send batches over the channel instead of individual items? They are batched downstream however
+                            foreach (var change in changes)
+                            {
+                                await channel.Writer.WriteAsync(change, cancellationToken).ConfigureAwait(false);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    Task IDashboardClient.WhenConnected => _whenConnected.Task;
```

Comment:
nit: WhenConnected -> Connected.

That's more in line with https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.hosting.ihostapplicationlifetime?view=dotnet-plat-ext-8.0

Reply:
Those are cancellation tokens. This is more in line with Task.WhenAny etc.

Reply:
I agree with James on the naming here. If you want another comparison channels has a completion task. Background service has one as well